sound/soc/sof/intel/hda-ipc.c

```diff
@@ -104,7 +104,9 @@ void hda_dsp_ipc_get_reply(struct snd_sof_dev *sdev)
 		ret = reply.error;
 	} else {
 		/* reply correct size ? */
-		if (reply.hdr.size != msg->reply_size) {
+		if (reply.hdr.size != msg->reply_size &&
+			/* getter payload is never known upfront */
+			!(reply.hdr.cmd & SOF_IPC_GLB_PROBE)) {
```

Comment:
@crojewsk shouldnt the check for cmd be first before you check for size?

Reply:
When it comes to optimization, one should validate the density of requests coming from each user group. In this case, it comes down to IPC-message calls from SOF handlers. Probe-IPCs ain't exactly the high tide here, these are few in numbers, really. Standard stream-ipcs run them over when it comes to numbers.

So, from framework's perspective it's _optimal_ to perform the size check first to address commons (as the evaluation to TRUE is also unexpected and should not happen in general) and only move to second instruction (cmd check) when necessary.

Reply:
Sorry, I don't see what optimization is taking place here.

Why would the probe be any different than other IPCs anyways? Put differently, is this a feature that the probe IPC has a reply which is different, and if yes why?

Reply:
Ranjani asked about the order of the operations, not about why the additional logical condition has been added.
The reasoning behind said statement though, is again connected to lack of GETs - topic which I have explained a lot already. Without this "hack", every single GET fails on kernel side.

Reply:
No problem, will clarify this once again.

Existing SOF IPC framework shares some issues from its older brother, /skylake (/common actually) IPC framework. One of which are getter-payloads. You've been reviewing my changes on alsa-devel some time ago, so I hope you understand the problem behind and why those patches where necessary.

Now in SOF, we are again standing in front of the exact same wall - this time we are not allowed to properly handle getters because IPC framework blocks ALL ipcs with payload of unexpected size. When we speak of any variable getter - which there are plenty of in cAVS, and here DMA_GET/ PROBE_POINTS_GET in SOF - host is never aware of returned size upfront and this is expected to allocate maximum buffer and await response from DSP. Once response is received, it parses data-size field (data_offset_size for cAVS, ipc::hdr::size for SOF) and adjusts (reallocates) the payload buffer accordingly.

Fixing this problem for SOF IPC is not in scope for Probes, so I've decided to run with this hack instead. Hope this clarifies the test now.